src/libraries/System.Private.CoreLib/src/System/Collections/Generic/Dictionary.cs

```diff
@@ -133,7 +133,15 @@ private void AddRange(IEnumerable<KeyValuePair<TKey, TValue>> collection)
                 if (source._comparer == _comparer)
                 {
                     // If comparers are the same, we can copy _entries without rehashing.
-                    CopyEntries(oldEntries, source._count);
+                    if (source._entries.Length == _entries.Length)
```

Comment:
This optimization was discussed and rejected in #41944.

Do you have data that prove that this situation common enough to justify adding this extra code that will make the Dictionary more expensive for everybody?

Reply:
The performance difference decreased between a straight copy as the dictionary increased in size (so at 3000 elements it was only x1.2 faster); when it goes over `BulkMoveWithWriteBarrierChunk` threshold of 16kB and starts batching calls to the extren copy so as not to starve the GC, when some of the batch advantages that the smaller sizes give (`xmm` widths, less writebarriers) start to diminish with the extra overhead https://github.com/dotnet/runtime/pull/41944#discussion_r487661992

Reply:
Hey Jan, in my teams repo we are often making a copy of very large dictionaries adding/updating a small number of items and then Interlocked.Exchange vs. the significant overhead of ConcurrentDictionary. Obviously this just one use case and I don't have data about how common this use case would be for the rest of the community.

Reply:
Hey @benaadams, I'm interested in you seeing the performance difference decrease as it seems to hold all the way to 2.5 million entries.
|              Method | NetFX |         N |            Mean |           Error |          StdDev |          Median |             Min |             Max | Ratio | RatioSD |  Allocated |
|-------------------- |------ |---------- |----------------:|----------------:|----------------:|----------------:|----------------:|----------------:|------:|--------:|-----------:|
| CopyConstructorFull |    PR |        10 |        121.7 ns |         7.87 ns |         9.07 ns |        118.2 ns |        111.4 ns |        140.0 ns |  1.00 |    0.00 |      440 B |
| CopyConstructorFull |  Base |        10 |        157.4 ns |         8.14 ns |         9.05 ns |        154.7 ns |        147.8 ns |        181.2 ns |  1.30 |    0.13 |      440 B |
|                     |       |           |                 |                 |                 |                 |                 |                 |       |         |            |
| CopyConstructorFull |    PR |       200 |        949.4 ns |        46.24 ns |        51.39 ns |        938.2 ns |        879.3 ns |      1,072.3 ns |  1.00 |    0.00 |     6824 B |
| CopyConstructorFull |  Base |       200 |      2,197.5 ns |        98.34 ns |       113.25 ns |      2,190.5 ns |      1,991.1 ns |      2,433.1 ns |  2.33 |    0.18 |     6824 B |
|                     |       |           |                 |                 |                 |                 |                 |                 |       |         |            |
| CopyConstructorFull |    PR |     3,000 |     12,534.2 ns |       371.41 ns |       427.71 ns |     12,644.8 ns |     11,738.0 ns |     13,079.9 ns |  1.00 |    0.00 |    94520 B |
| CopyConstructorFull |  Base |     3,000 |     31,799.4 ns |     2,369.86 ns |     2,729.14 ns |     31,739.8 ns |     27,901.2 ns |     36,866.5 ns |  2.54 |    0.23 |    94520 B |
|                     |       |           |                 |                 |                 |                 |                 |                 |       |         |            |
| CopyConstructorFull |    PR |    50,000 |    565,528.3 ns |    66,711.77 ns |    76,825.38 ns |    573,232.5 ns |    413,883.9 ns |    677,965.5 ns |  1.00 |    0.00 |  1466305 B |
| CopyConstructorFull |  Base |    50,000 |    939,847.5 ns |    76,891.54 ns |    88,548.42 ns |    949,663.8 ns |    716,979.8 ns |  1,123,293.9 ns |  1.70 |    0.36 |  1466957 B |
|                     |       |           |                 |                 |                 |                 |                 |                 |       |         |            |
| CopyConstructorFull |    PR |   200,000 |  3,003,583.0 ns |   214,751.81 ns |   247,308.51 ns |  3,065,712.5 ns |  2,367,460.0 ns |  3,318,942.5 ns |  1.00 |    0.00 |  6309154 B |
| CopyConstructorFull |  Base |   200,000 |  5,016,504.1 ns |   319,545.72 ns |   367,989.32 ns |  4,963,555.2 ns |  4,363,316.7 ns |  5,856,329.2 ns |  1.68 |    0.18 |  6309403 B |
|                     |       |           |                 |                 |                 |                 |                 |                 |       |         |            |
| CopyConstructorFull |    PR |   500,000 |  5,802,838.2 ns |   293,421.07 ns |   313,957.26 ns |  5,696,210.9 ns |  5,418,712.5 ns |  6,466,756.2 ns |  1.00 |    0.00 | 15699660 B |
| CopyConstructorFull |  Base |   500,000 | 13,841,580.6 ns |   974,592.43 ns | 1,083,256.90 ns | 13,458,152.6 ns | 12,760,700.0 ns | 16,475,305.3 ns |  2.39 |    0.21 | 15699848 B |
|                     |       |           |                 |                 |                 |                 |                 |                 |       |         |            |
| CopyConstructorFull |    PR | 1,000,000 | 14,965,381.2 ns |   633,808.01 ns |   704,475.92 ns | 14,978,606.2 ns | 13,380,162.5 ns | 16,363,681.2 ns |  1.00 |    0.00 | 32555616 B |
| CopyConstructorFull |  Base | 1,000,000 | 39,526,627.5 ns | 4,969,687.48 ns | 5,723,099.50 ns | 39,550,675.0 ns | 29,117,750.0 ns | 49,635,616.7 ns |  2.67 |    0.38 | 32555661 B |
|                     |       |           |                 |                 |                 |                 |                 |                 |       |         |            |
| CopyConstructorFull |    PR | 2,500,000 | 33,532,232.1 ns | 2,965,513.86 ns | 3,415,090.17 ns | 34,534,078.6 ns | 29,052,414.3 ns | 38,902,800.0 ns |  1.00 |    0.00 | 81012789 B |
| CopyConstructorFull |  Base | 2,500,000 | 79,514,458.8 ns | 5,650,372.56 ns | 6,506,977.45 ns | 79,375,100.0 ns | 69,956,350.0 ns | 92,912,075.0 ns |  2.40 |    0.36 | 81011780 B |

Reply:
Was a `string` key; so it invoked the GC write barrier

Reply:
Ah, I see so this optimization only performs for reference type keys which would mean that this is more of an edge case for applicability. @jkotas let me know if you'd like to abandon this change.

Reply:
Yes, I think this change should be abandoned. Thank you for giving it a shot anyway!